third_party/move/move-vm/runtime/src/loader/mod.rs

```diff
@@ -1957,44 +1956,38 @@ impl Module {
                     instantiation.push(cache.make_type_while_loading(&module, ty)?);
                 }
                 struct_instantiations.push(StructInstantiation {
-                field_count,
+                    field_count,
                     instantiation,
                     definition_struct_type: struct_def.definition_struct_type.clone(),
                 });
             }
 
-            for func_handle in module.function_handles() {
-                let func_name = module.identifier_at(func_handle.name);
-                let module_handle = module.module_handle_at(func_handle.module);
-                let module_id = module.module_id_for_handle(module_handle);
-                if module_id == id {
-                    // module has not been published yet, loop through the functions
-                    for (idx, function) in cache.functions.iter().enumerate().rev() {
-                        if function.module_id() != Some(&module_id) {
-                            return Err(PartialVMError::new(
-                                StatusCode::FUNCTION_RESOLUTION_FAILURE,
-                            )
-                            .with_message(format!(
-                                "Cannot find {:?}::{:?} in publishing module",
-                                module_id, func_name
-                            )));
-                        }
-                        if function.name.as_ident_str() == func_name {
-                            function_refs.push(idx);
-                            break;
-                        }
-                    }
-                } else {
-                    function_refs.push(cache.resolve_function_by_name(func_name, &module_id)?);
-                }
-            }
+            for (idx, func) in module.function_defs().iter().enumerate() {
+                let findex = FunctionDefinitionIndex(idx as TableIndex);
+                let mut function = Function::new(natives, findex, func, &module);
+                function.return_types = function
```

Comment:
Why is this code now suddenly here? Can we add a comment? is this logic tested?
I see that the loader in add_module used to have these exact same code fragment. Is it still there? Can the code be re-used?

Also looking at make_type_while_loading implementation, each call literally linearly traverses a list of structs, we can at least make a batch interface? i.e. collect all toks here in some data-structure, and make_types_while_loading.
Similarly elsewhere. Or maybe the structs in module can be a more efficient data-structure? 

@runtian-zhou @ziaptos can that be costly? it's only for publishing, and I suppose the amount of structs is bounded?

Reply:
I guess the chunk I found was removed in the next commit, I think some of these commits could have been squashed, but it's fine. The efficiency question is still there for me though, and also why moving it from add_module to here is equivalent.

Reply:
This was from a previous review request from Zekun. Let's talk offline about the concern here cuz I wasn't really certain what's your concern over the code efficiency.

Reply:
This wouldn't be a problem for this PR anyway, but make_type_while_loading seems to be doing a linear scan each time, was the thing.